ads/google/a4a/line-delimited-response-handler.js

```diff
@@ -0,0 +1,164 @@
+  * @param {?../../../src/service/xhr-impl.FetchInitDef=} opt_init
+  * @return {!Promise<!../../../src/service/xhr-impl.FetchResponse>} response,
+  *    note that accessing body will cause error as its consumed in streaming.
+  */
+ export function fetchLineDelimitedChunks(xhr, url, chunkHandler, opt_init) {
+   return xhr.fetch(url, opt_init)
+     .then(response => {
+       if (!response.body || !xhr.win.TextDecoder) {
+         // TODO(keithwrightbos) - TextDecoder polyfill?
+         response.text().then(content =>
+           chunkHandleFullResponse(content, chunkHandler));
+       } else {
+         let metaData;
+         let snippet = '';
+         const chunkCallback = (chunk, done) => {
+           let hasDelimiter;
+           let pos = 0;
+           do {
+             const delimIndex = chunk.indexOf('\n', pos);
+             hasDelimiter = delimIndex != -1;
+             let end = hasDelimiter ? delimIndex : chunk.length;
+             snippet += chunk.substr(pos, end - pos);
+             pos = hasDelimiter ? delimIndex + 1 : end;
+             if (hasDelimiter) {
+               if (!metaData) {
+                 metaData = safeJsonParse_(snippet);
+               } else {
+                 // Have metaData/HTML so we can call chunkHandler and reset.
+                 chunkHandler(
+                   unescapeHtml_(snippet),
+                   /** @type {!Object<string, *>} */(metaData),
+                   done && end == chunk.length);
+                 metaData = undefined;
+               }
+               snippet = '';
+             } else if (countTrailingSlashes(chunk, pos - 1) & 1) {
+               // Decrement 'end' to skip the last character of the chunk if it
+               // ends with an odd number of slashes, since that implies that
+               // the last slash is used to escape a character that we haven't
+               // received yet. So we skip the last byte in this iteration and
+               // wait for the next chunk to see what is being escaped.
+               end--;
+             }
+             pos = hasDelimiter ? delimIndex + 1 : end;
+           } while (hasDelimiter && pos < chunk.length);
+         };
+         handleFetchResponseStream_(
+           response.body.getReader(), chunkCallback, new TextDecoder('utf-8'));
+       }
+       return response;
+     });
+ }
+
+ /**
+  * @param {!ReadableStreamReader} reader The reader of the fetch's
+  *  response stream.
+  * @param {!function(string, boolean)} chunkHandler
+  * @param {!TextDecoder} textDecoder
+  * @private
+  */
+ function handleFetchResponseStream_(reader, chunkHandler, textDecoder) {
+   return reader.read().then(result => {
+     if (result.value) {
+       chunkHandler(
+         textDecoder.decode(
+           /** @type {!ArrayBuffer} */(result.value), {'stream': true}),
+         result.done);
+     }
+     if (!result.done) {
+       // More chunks to read.
+       handleFetchResponseStream_.apply(this, arguments);
```

Comment:
This is not very efficient, it'll recreate a closure every time (you'll face an `arguments` usage deoptimization). Instead:

```js
handleFetchStream(reader, handle, decoder) {
  reader.read().then(function chunk(result) {
    // handle
    if (!result.done) {
      reader.read().then(chunk);
    }
  });
}